### PR TITLE
[SYCL][ABI-Break] Re-design the handling of optional code location argument for USM calls and honor memory allocation properties

### PR DESCRIPTION
**Pull request by smaslov-intel**
Do not drop properties specified by user.

Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

### CODE REVIEW

**Comment by bader** [comment]
This fixes just one function out of eight. The properties are dropped by `malloc`, `malloc_device`, `malloc_host`, `malloc_shared`, `aligned_alloc_device`, `aligned_alloc_host`, `aligned_alloc_shared`. Shouldn't we fix all of them?

**Comment by smaslov-intel** [comment]
> This fixes just one function out of eight. The properties are dropped by `malloc`, `malloc_device`, `malloc_host`, `malloc_shared`, `aligned_alloc_device`, `aligned_alloc_host`, `aligned_alloc_shared`. Shouldn't we fix all of them?

We probably should. I fixed where I saw the immediate problem reported. Do you prefer we do all at once in this PR?

**Comment by smaslov-intel** [comment]
BTW, my change broke ABI test (due to new argument added). 

> c++filt
The following symbols are missing from the new object file:
_ZN2cl4sycl13aligned_allocEmmRKNS0_6deviceERKNS0_7contextENS0_3usm5allocENS0_6detail13code_locationE

I guess I should just leave the existing version and add one that accepts properties, right?
Or, maybe we are allowed to break ABI now?

**Comment by aelovikov-intel** [comment]
> Or, maybe we are allowed to break ABI now?

We are allowed.

**Comment by bader** [comment]
> > This fixes just one function out of eight. The properties are dropped by `malloc`, `malloc_device`, `malloc_host`, `malloc_shared`, `aligned_alloc_device`, `aligned_alloc_host`, `aligned_alloc_shared`. Shouldn't we fix all of them?
> 
> We probably should. I fixed where I saw the immediate problem reported. Do you prefer we do all at once in this PR?

Yes. It's better to fix all them at once and don't wait 7 other bug reports from customers to fix. :)

**Comment by smaslov-intel** [comment]
> > It's better to fix all them at once
> 
> @bader, sure, it would just cause this customer problem to hang for some more time. I now find quite a few duplicates of the alloc routines seeming to do with an extra argument, which is not in the spec:
> 
> > const detail::code_location CL
> 
> Any idea where are these coming from and which should be removed, if any?

It looks they are coming from https://github.com/intel/llvm/pull/5391
But why do we have version with "code_location" and without it?
Which one should stay? (SYCL spec doesn't know about code_location, which is from "detail")

**Comment by bader** [comment]
I don't know for sure, but I speculate that it's added to support the tools added with https://github.com/intel/llvm/pull/5389 and `sycl-sanitize` specifically.
I hope someone from @intel/llvm-reviewers-runtime team and provide more details about that. Unfortunately, all folks working on this are not available.

**Comment by smaslov-intel** [comment]
> I don't know for sure, but I speculate that it's added to support the tools added with #5389 and `sycl-sanitize` specifically. I hope someone from @intel/llvm-reviewers-runtime team and provide more details about that. Unfortunately, all folks working on this are not available.

Thanks! I am specifically worried that we are polluting the "sycl" namespace with something that is not a part of the SYCL spec.

**Comment by bader** [comment]
> > I don't know for sure, but I speculate that it's added to support the tools added with #5389 and `sycl-sanitize` specifically. I hope someone from @intel/llvm-reviewers-runtime team and provide more details about that. Unfortunately, all folks working on this are not available.
> 
> Thanks! I am specifically worried that we are polluting the "sycl" namespace with something that is not a part of the SYCL spec.

I remember that this approach was originally added by @tovinkere to enable tools support. https://github.com/intel/llvm/pull/1129

I vaguely recall discussing this concern with him and he convinced me that there should be no impact on users.

**Comment by smaslov-intel** [comment]
> I remember that this approach was originally added by @tovinkere to enable tools support. #1129

Thanks @bader ! I discussed this with @tovinkere and uploaded a patch that I think does what is wanted.
It follows https://github.com/intel/llvm/pull/1129 and allows to disable the optional last `code_location` argument with DISABLE_SYCL_INSTRUMENTATION_METADATA.

**Review by bader — COMMENTED** [review]
Please, update declarations in the header file - sycl/include/sycl/usm.hpp.

**Comment by smaslov-intel** [comment]
> Please, update declarations in the header file - sycl/include/sycl/usm.hpp.

Done

File: sycl/include/sycl/usm.hpp
Comment (bader) [inline sycl/include/sycl/usm.hpp:28]: Shouldn't we remove (or deprecate) overloads w/o property_list? The spec declarations have property_list parameter.

File: sycl/include/sycl/usm.hpp
Comment (bader) [inline sycl/include/sycl/usm.hpp:293]: I think template versions must use `#ifndef DISABLE_SYCL_INSTRUMENTATION_METADATA` too. Right?

File: sycl/include/sycl/usm.hpp
Comment (smaslov-intel) [reply sycl/include/sycl/usm.hpp:28]: > Shouldn't we remove (or deprecate) overloads w/o property_list? The spec declarations have property_list parameter.

I wish we could do that, and I actually removed them initially.
But they are needed due to how FE instruments the calls, it adds the code_location after last given argument before the default initialized property_list is resolved, so I got lots of fails of unresolved symbols (where propery_list wasn't on the call).

File: sycl/include/sycl/usm.hpp
Comment (smaslov-intel) [reply sycl/include/sycl/usm.hpp]: I think so too, PR was just growing too large. Updated it.

Comment (bader) [reply sycl/include/sycl/usm.hpp]: I'm okay to separate this change form this PR if it helps.

**Comment by smaslov-intel** [comment]
The fails in `SYCL / Linux / ESIMD Emu LLVM Test Suite (pull_request)` are not related to this PR:

> Failed Tests (3):
  SYCL :: dpas/dpas_test1.cpp
  SYCL :: dpas/dpas_test2.cpp
  SYCL :: dpas/dpas_test3.cpp

**Review by bader — DISMISSED** [review]
👍 Thanks!

**Review by tovinkere — CHANGES_REQUESTED** [review]
Moving _CODELOCPARAM and other macros to common.hpp and simplifying the default argument and its use in the implementation will make it more readable and define a template for use in other places in the future.

File: sycl/include/sycl/usm.hpp
Comment (smaslov-intel) [reply sycl/include/sycl/usm.hpp:300]: I do like C macros but not when they hide "," or require anything from following code, like you show.

File: sycl/include/sycl/usm.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/usm.hpp:27]: Can we abstract the whole thing into a single macro? Something like https://godbolt.org/z/ojG9zc81z

File: sycl/include/sycl/usm.hpp
Comment (aelovikov-intel) [reply sycl/include/sycl/usm.hpp:300]: I've missed @tovinkere 's comment while writing my response, but I do think that his approach is way nicer and easier to maintain than current version. All those unnecessary preprocessor `if`s are so hard to read and they just steal attention from what's really meaningful in the code.

Comment (aelovikov-intel) [reply sycl/include/sycl/usm.hpp:300]: > require anything from following code

We can avoid that by standardizing the name of the code location and declaring it as `inline constexpr` variable in the namespace scope.

Comment (tovinkere) [reply sycl/include/sycl/usm.hpp:300]: > I do like C macros but not when they hide "," or require anything from following code, like you show.

I am in general agreement with your position, but this use requires an #ifndef and #ifdef pair that makes it easy for someone to make a mistake that may go unnoticed. Using these macros is one way to prevent mistakes and have consistency.

Comment (smaslov-intel) [reply sycl/include/sycl/usm.hpp]: I've uploaded a patch that uses macros, which you'd hopefully like better.

**Comment by smaslov-intel** [comment]
The ESIMD and CUDA failures are unrelated.

File: sycl/include/sycl/usm.hpp
Comment (aelovikov-intel) [reply sycl/include/sycl/usm.hpp]: I think this PR should be split in two. The refactoring you've done for handling code location is rather significant.

Comment (smaslov-intel) [reply sycl/include/sycl/usm.hpp]: > I think this PR should be split in two. The refactoring you've done for handling code location is rather significant.

I've started with a simple change where properties were not dropped, but other reviewers requested a more complete change. At this point I do not see a benefit of splitting this PR. Just pretend that this refactoring is the goal of this PR although it also fixes some small functional issue.

**Review by steffenlarsen — COMMENTED** [review]
Looks like it breaks ABI so the title should reflect that and the ABI version should be incremented.

Also, before it gets merged it may be worth checking the windows dump and making the corresponding changes there as well so we avoid post-commit failures.

File: sycl/include/sycl/usm.hpp
Comment (tovinkere) [reply sycl/include/sycl/usm.hpp]: > > I think this PR should be split in two. The refactoring you've done for handling code location is rather significant.
> 
> I've started with a simple change where properties were not dropped, but other reviewers requested a more complete change. At this point I do not see a benefit of splitting this PR. Just pretend that this refactoring is the goal of this PR although it also fixes some small functional issue.

I am okay with not having two PRs for this - the one thing I do not like is that queue.hpp and usm.hpp implement two different sets of macros. I would hate to have a PR just to reconcile the different approaches.

Comment (smaslov-intel) [reply sycl/include/sycl/usm.hpp]: These aren't functional macros but rather some convenience ones (and quite simple) that some **may** prefer using over more explicit coding. I can imagine (and support) people creating a third variant if they like it better. Whose final call is it how to proceed with this PR?

@steffenlarsen : I've change the title to reflect it is an ABI breaking change, but still having trouble to apply/test these changes on Windows. Following up.

**Comment by smaslov-intel** [comment]
> Also, before it gets merged it may be worth checking the windows dump and making the corresponding changes there as well so we avoid post-commit failures.

@steffenlarsen : I am getting strange errors when trying to configure the build:

> $ python buildbot/configure.py -t Debug
...
-- Looking for _M_X64
CMake Error: Unable to open check cache file for write. 

Is there a way to trigger Win build in CI/pre-commit?

File: sycl/include/sycl/usm.hpp
Comment (smaslov-intel) [reply sycl/include/sycl/usm.hpp]: @tovinkere : I thought there is no reason for me to get back to it, so I pushed a change that would remain you the principal owner of the method :)

**Review by tovinkere — APPROVED** [review]
LGTM.

**Comment by smaslov-intel** [comment]
> Also, before it gets merged it may be worth checking the windows dump and making the corresponding changes there as well so we avoid post-commit failures.

I have updated the windows symbols. Please review/approve/merge.

**Comment by smaslov-intel** [comment]
@aelovikov-intel , @bader , @steffenlarsen , @tovinkere : ping
The ABI breaking changes get outdate quickly this time and it's far from friendly to get symbols updated

File: sycl/include/sycl/usm.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/usm.hpp:36]: nit: I don't think reference `&` makes any change, neither for `&CL` nor for `&propList` below. It's also harmless though.

File: sycl/include/sycl/detail/common.hpp
Comment (aelovikov-intel) [inline sycl/include/sycl/detail/common.hpp:87]: nit: I don't think FE does anything here, the solution is purely on the side of the library using common builtins.

File: sycl/source/detail/usm/usm_impl.cpp
Comment (aelovikov-intel) [inline sycl/source/detail/usm/usm_impl.cpp:435]: Why do we need copy-paste in this overload and the two below. Can't we route all them to a single implementation similarly to left side?

File: sycl/include/sycl/detail/common.hpp
Comment (smaslov-intel) [reply sycl/include/sycl/detail/common.hpp:87]: it is FE that either adds the code_location parameter or not

File: sycl/source/detail/usm/usm_impl.cpp
Comment (smaslov-intel) [reply sycl/source/detail/usm/usm_impl.cpp:435]: I don't recall exactly now, but there were "ambiguous" definitions issues at some point. These are small enough to get inlined.

**Review by bader — APPROVED** [review]

File: sycl/source/detail/usm/usm_impl.cpp
Comment (steffenlarsen) [inline sycl/source/detail/usm/usm_impl.cpp:448]:
```suggestion
  if (Kind == alloc::host)
    return detail::usm::alignedAllocHost(0, Size, Ctxt, Kind, property_list{},
                                         CL);
  return detail::usm::alignedAlloc(0, Size, Ctxt, Dev, Kind, property_list{},
                                   CL);
```
Same applies to the next couple functions.

File: sycl/include/sycl/usm.hpp
Comment (steffenlarsen) [inline sycl/include/sycl/usm.hpp:24]: A bit of a nitpick but arguably `CodeLoc` was a better name as it better conveys the intention and hopefully has less chance to conflict with macros.
I realize we had a bunch of `CL` parameters before so it is not make-or-break, but we might as well improve it while we are in here. 😄

Comment (smaslov-intel) [reply sycl/include/sycl/usm.hpp:24]: Thanks for feedback but I think `CL` is sufficiently wrapped with more descriptive macros like `_CODELOCPARAM` to be a NIT. Do you agree?

Comment (steffenlarsen) [reply sycl/include/sycl/usm.hpp:24]: Good point. I suppose it is, except for the cases where they are passed on. That only leaves the macro collision problem, but in the end that is already a topic being discussed, so as long as `CL` isn't a commonly-used macro I am okay with it for now.

File: sycl/source/detail/usm/usm_impl.cpp
Comment (steffenlarsen) [reply sycl/source/detail/usm/usm_impl.cpp]: Note: There are more functions like this. I only made the suggestion for the first one, but the changes should be applied to all of them.

**Review by steffenlarsen — APPROVED** [review]
LGTM! 😄

**Comment by steffenlarsen** [comment]
> Anything prevents this from being merged?

I don't think so. Merging!

**Pull request merged**
